class.c

```diff
@@ -63,61 +68,96 @@ rb_module_add_to_subclasses_list(VALUE module, VALUE iclass)
 {
     rb_subclass_entry_t *entry, *head;
 
-    entry = ALLOC(rb_subclass_entry_t);
+    entry = ZALLOC(rb_subclass_entry_t);
     entry->klass = iclass;
-    entry->next = NULL;
 
     head = RCLASS_SUBCLASSES(module);
-    if (head) {
-	entry->next = head;
-	RCLASS_MODULE_SUBCLASSES(head->klass) = &entry->next;
+    if (!head) {
+        head = ZALLOC(rb_subclass_entry_t);
+        RCLASS_SUBCLASSES(module) = head;
+    }
+    entry->next = head->next;
+    entry->prev = head;
+
+    if (head->next) {
+        head->next->prev = entry;
     }
+    head->next = entry;
```

Comment:
I think this code can be extracted into a function since it's the same as the one in `rb_class_subclass_add`.